### PR TITLE
How does this inotify exporter look?

### PR DESCRIPTION
**Pull request by stephen-soltesz**
This PR is for feedback before writing more code or tests.

What would you change about the way that the logic is structured?

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/inotify-exporter/1)

### CODE REVIEW

File: cmd/inotify_exporter/main.go [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+	dayPattern   = regexp.MustCompile("0[1-9]|[12][0-9]|3[0-1]")
+
+	// activeWatches prevents watching the same path twice.
+	activeWatches = &watches{make(map[string]chan notify.EventInfo), sync.Mutex{}}
+)
+
+// Prometheus Metrics.
+var (
+	createCount = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "inotify_file_create_total",
```
Comment (gfr10598) [inline cmd/inotify_exporter/main.go:56]: Drop _total from the names?
Also, is "file" redundant?  Will we ever count something other than file events?

File: cmd/inotify_exporter/main.go [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+	// createExtensions.WithLabelValues("s2c_snaplog").Inc()
+	createExtensions = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "inotify_extension_create_total",
+			Help: "A running count of file extensions created in watched directories.",
+		},
+		[]string{"ext"},
+	)
+	deleteCount = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "inotify_file_delete_total",
```
Comment (gfr10598) [inline cmd/inotify_exporter/main.go:68]: Is there any value add to this?  Seems like it is redundant with the accounting by extension.

File: cmd/inotify_exporter/main.go [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/rjeczalik/notify"
```
Comment (gfr10598) [inline cmd/inotify_exporter/main.go:32]: Depending on how stable this is, we may want to vendor cache it?

**Review by gfr10598 — COMMENTED** [review]
Overall really like it!
I've left a few comments, but will defer a careful review until you tell me it is stable.
Maybe not too early to add some unit tests?

**Comment by stephen-soltesz** [comment]
It is definitely time to add unit tests. I wanted to make sure the rest looked right before taking more time to do that in case it turned out to require major changes.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[cmd/inotify_exporter/main.go, line 32 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-KwLThqudPqwvtAR-p0D-r3-32:-KwLW8ZCkRW4b_gI_rhT:beo7vqj) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L32)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Depending on how stable this is, we may want to vendor cache it?
</blockquote></details>

Yes, funny dependencies like this would probably be safer to vendor.

---

*[cmd/inotify_exporter/main.go, line 56 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-KwLThqudPqwvtAR-p0D-r3-56:-KwLV4WP7o1Z7SyFZaIL:bef1fw2) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L56)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Drop _total from the names?
Also, is "file" redundant?  Will we ever count something other than file events?
</blockquote></details>

For better or worse the `_total` suffix is recommended best practice by prometheus.
https://prometheus.io/docs/practices/naming/

The `promtool` metrics linter reports an error if a counter does not have that suffix.

```
curl http://localhost:9494/metrics | ~/bin/promtool check-metrics
inotify_extension_create: counter metrics should have "_total" suffix
...
```

My thought is that "file" in the metric name is not redundant because the inotify mechanism can also report events on directories. My thought is that the name helps someone unfamiliar with the implementation interpret the meaning of the metric without deeper knowledge. Hopefully one day the user of metrics in prometheus won't necessarily be the author of the tools generating them.

---

*[cmd/inotify_exporter/main.go, line 68 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-KwLThqudPqwvtAR-p0D-r3-68:-KwLgzrNDlnVIpuHV1Bu:b672crz) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L68)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Is there any value add to this?  Seems like it is redundant with the accounting by extension.
</blockquote></details>

My hope was that `inotify_file_create_total` - `inotify_file_delete_total` would equal the total number of files on disk. That's true if everything starts from zero, but if the exporter is restarted while there are files on disk, then we will end up counting more deletes than creates and the difference is misleading.

There doesn't seem to be a good way of estimating or reading the number of files in a directory with out explicitly listing and counting them.

So, my thinking is that even if the "create - delete" difference is not valuable, that being able to see the rate of deletes is helpful because this is a sign that delete logs safely is operating (or whatever issues deletes in the future). And correlating this value with the disk io stats could identify the relative contribution of overhead or improvements when we use new strategies.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-:-KxJflj_e8GhgPX9kQKA:b-3bgme0)*

**Comment by gfr10598** [comment]
*[cmd/inotify_exporter/main.go, line 68 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-KwLThqudPqwvtAR-p0D-r3-68:-Kxn6bBV6sUrFqEyEDvg:bevkwa5) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L68)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

My hope was that `inotify_file_create_total` - `inotify_file_delete_total` would equal the total number of files on disk. That's true if everything starts from zero, but if the exporter is restarted while there are files on disk, then we will end up counting more deletes than creates and the difference is misleading.

There doesn't seem to be a good way of estimating or reading the number of files in a directory with out explicitly listing and counting them.

So, my thinking is that even if the "create - delete" difference is not valuable, that being able to see the rate of deletes is helpful because this is a sign that delete logs safely is operating (or whatever issues deletes in the future). And correlating this value with the disk io stats could identify the relative contribution of overhead or improvements when we use new strategies.
</blockquote></details>

Ah - what I mean is that deleteCount is redundant with deleteExtensions.  Isn't everything counted here also counted there?  Same for createXXX.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1)*

**Comment by stephen-soltesz** [comment]
I'm going to close this trial PR in favor of -- https://github.com/m-lab/inotify-exporter/pull/2

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[cmd/inotify_exporter/main.go, line 68 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-KwLThqudPqwvtAR-p0D-r3-68:-KxnAg6hK3YsMRsEn0-9:bmi1zu8) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L68)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Ah - what I mean is that deleteCount is redundant with deleteExtensions.  Isn't everything counted here also counted there?  Same for createXXX.
</blockquote></details>

Ooh, yes actually they are redundant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-:-KxnAmxMfbEuw--DBmeb:b-mxrctc)*

**Comment by stephen-soltesz** [comment]
Closing in favor of https://github.com/m-lab/inotify-exporter/pull/2

**Comment by gfr10598** [comment]
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions.

---

*[cmd/inotify_exporter/main.go, line 32 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-KwLThqudPqwvtAR-p0D-r3-32:-Kxn5w4h5IOufXT4Avkj:b6pdot) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L32)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Yes, funny dependencies like this would probably be safer to vendor.
</blockquote></details>

Add a TODO to consider vendoring.

---

*[cmd/inotify_exporter/main.go, line 241 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-Kxn9lpDHr2qDof5ipLg:-Kxn9lpDHr2qDof5ipLh:bv4nn1z) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L241)):*
> ```Go
> // with the time of the event, the event info, and a logger instance.  The
> // watch will remain active until dayDir is deleted.
> func watchDay(dayDir string, startTime time.Time,
> ```

I notice this signature swaps the dayDir and startTime compared to all the other watchXXX functions.  Is that intentional?

---

*[cmd/inotify_exporter/main.go, line 241 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-KxnARhqjx8JtRbO89lj:-KxnARhqjx8JtRbO89lk:bwkuuzm) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L241)):*
> ```Go
> // with the time of the event, the event info, and a logger instance.  The
> // watch will remain active until dayDir is deleted.
> func watchDay(dayDir string, startTime time.Time,
> ```

Please document what startTime is for.

---

*[cmd/inotify_exporter/main.go, line 268 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-KxnAuqmNCQ3XORm6sf8:-KxnAuqmNCQ3XORm6sf9:bhrliy9) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L268)):*
> ```Go
> 
> 	// Extract the directory prefix and the YYYY/MM/DD suffix from dayDir.
> 	yearMonthDay := dayDir[len(dayDir)-10:]
> ```

How about using a regexp, here and other places?  Might make the isValidXXX irrelevant as well?

---

*[cmd/inotify_exporter/main.go, line 414 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-Kxn8sOEx6RAFLtSFJQL:-Kxn8sOEx6RAFLtSFJQM:b3r8l0o) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L414)):*
> ```Go
> }
> 
> func watchCurrentDay(start time.Time, dayDir string) {
> ```

watchDay.  Am I missing something, or is this an arbitrary day, but you will be passing in today?

---

*[cmd/inotify_exporter/main.go, line 467 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-Kxn8UM9N1I4kdRR6tyr:-Kxn8UM9N1I4kdRR6tys:b-8rrx2m) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L467)):*
> ```Go
> func main() {
> 	flag.Parse()
> 	// Watch root dir.
> ```

This feels like both an outline, and documentation.  It will evolve into the latter?

---

*[cmd/inotify_exporter/main.go, line 480 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-Kxn7azuvf3PZfjxKwkc:-Kxn7azuvf3PZfjxKwkd:b-7ilkmm) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L480)):*
> ```Go
> 	//
> 	// So, at all times, there should be at least:
> 	// * one year watcher.
> ```

* one root watcher ?

---

*[cmd/inotify_exporter/main.go, line 486 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-Kxn7pgqapD4V29W6Nv_:-Kxn7pgqapD4V29W6Nva:b-6zczx6) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L486)):*
> ```Go
> 	// Each watcher will have a date associated with it. When an event occurs,
> 	// it can check the date. If the date is in the future, we can stop
> 	// watching.
> ```

Not sure about this.  Experiment could open a file with a date in the last minutes of yesterday, but the action time is today.  Also, the deletes will happen hours later.  So perhaps we should continue to watch until the directory is empty or deleted?

---

*[cmd/inotify_exporter/main.go, line 494 at r3](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1#-Kxn8Iaavgxt3ATv1PFV:-Kxn8Iaavgxt3ATv1PFW:b8mrecp) ([raw file](https://github.com/m-lab/inotify-exporter/blob/fffa493a00e449618565dc4aa591a52976cca706/cmd/inotify_exporter/main.go#L494)):*
> ```Go
> 	go func() {
> 		http.Handle("/metrics", promhttp.Handler())
> 		log.Fatal(http.ListenAndServe(":9494", nil))
> ```

Is this port meaningful or random, or temporary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/inotify-exporter/1)*

**Pull request closed without merging**